beard/metrics/text.py

```diff
@@ -0,0 +1,153 @@
+
+    Parameters
+    ----------
+    :param s: string
+        String to be searched
+
+    :param letter: string
+        Substring we are searching for
+
+    Returns
+    -------
+    :returns: generator
+        A generator that holds the indexes of the patterns
+    """
+    for match in re.finditer(pattern, s):
+        yield match.start()
+
+
+def _jaro_matching(s1, s2):
+    """Return the number of matching letters.
+
+    Calculate the number of matching letters
+    based on the method for computing the Jaro distance metric.
+
+    Parameters
+    ----------
+    :param s1: string
+        First string
+
+    :param s2: string
+        Second string
+
+    Returns
+    -------
+    :returns: int
+        The number of matching letters
+    """
+    H = min(len(s1), len(s2))//2
+
+    letters_cache = {}
+    matches = 0
+    transpositions = 0
+    s1_matching_letter_sequence = []
+    s2_matching_letter_sequence = []
+    s1_matched_positions = []
+    s2_matched_positions = []
+
+    for letter in s1:
+        if letter not in letters_cache:
+            letters_cache[letter] = (tuple(find_all(s1, letter)),
+                                     tuple(find_all(s2, letter)))
+
+    for letter, (s1_occur, s2_occur) in letters_cache.items():
+        for pos_s1 in s1_occur:
+            for pos_s2 in s2_occur:
+                if pos_s1 - H <= pos_s2 <= pos_s1 + H:
+                    if pos_s2 not in s2_matched_positions:
+                        matches += 1
+                        s2_matched_positions.append(pos_s2)
+                    s1_matching_letter_sequence.append((pos_s1, letter))
+                    break
+
+    for letter, (s1_occur, s2_occur) in letters_cache.items():
+        for pos_s2 in s2_occur:
+            for pos_s1 in s1_occur:
+                if pos_s1 - H <= pos_s2 <= pos_s1 + H:
+                    if pos_s1 not in s1_matched_positions:
+                        s1_matched_positions.append(pos_s1)
+                    s2_matching_letter_sequence.append((pos_s2, letter))
+                    break
+
+    s1_matching_letter_sequence.sort()
+    s2_matching_letter_sequence.sort()
+    transpositions = len(tuple(filter(lambda x: x[0][1] != x[1][1],
+                               zip(s1_matching_letter_sequence,
+                                   s2_matching_letter_sequence))))
+    return matches, transpositions
+
+
+def jaro(s1, s2):
+    """Return the Jaro similarity of the strings s1 and s2.
+
+    Parameters
+    ----------
+    :param s1: string
+        First string
+
+    :param s2: string
+        Second string
+
+    Returns
+    -------
+    :returns: float
+        Similarity of s1 and s2
+    """
+    n_matches, n_transpositions = _jaro_matching(s1, s2)
+    s = len(s1)
+    t = len(s2)
+
+    return (1/3)*((n_matches/s) +
+                  (n_matches/t) +
+                  ((n_matches - (n_transpositions/2))/(n_matches)))
+
+
+def jaro_winkler(s1, s2):
```

Comment:
I have seen several papers where they adjust the scaling factor. I think it is best to leave open the possibility to do that as well, as long as the default value matches the default algorithm.